src/RedisIPCMessenger.ts

```diff
@@ -0,0 +1,148 @@
+  }
+
+  async getOtherInstances(room: Room): Promise<Array<Instance>> {
+    this.validateRoom(room);
+    this.makeSureRoomIsJoined(room);
+
+    const keys = await this.publisher.keys(`${room}:*`);
+    return keys.map((key) => {
+      const [, instance] = key.split(':');
+      return instance;
+    }).filter((instance) => instance !== this.instance);
+  }
+
+  async send(room: Room, message: Omit<Message, 'sender'>): Promise<void> {
+    this.validateRoom(room);
+    this.makeSureRoomIsJoined(room);
+
+    await this.publisher.publish(room, this.serialize({
+      ...message,
+      sender: this.instance,
+    }));
+  }
+
+  private validateRoom(room: string) {
+    if (!room.length || room.includes(':')) {
+      throw new TypeError(`room must be a non empty string and should not include ':', ${room} received`);
+    }
+  }
+
+  private validateInstance(instance: string) {
+    if (!instance.length || instance.includes(':')) {
+      throw new TypeError(`room/instance must be a non empty string and should not include ':', ${instance} received`);
+    }
+  }
+
+  private startRefreshKeyLoop = async (room: Room) => {
+    const refreshKeyLoop = async () => {
+      await this.publisher.set(`${room}:${this.instance}`, '', 'EX', this.expireTime / 1000);
+      setTimeout(refreshKeyLoop, this.refreshInterval);
+    };
+
+    await refreshKeyLoop();
+  };
+
+  private storeSubscription(room: Room, callback: MessageCallback) {
+    this.subscriptions.set(room, callback);
+  }
+
+  private async subscribeForRoomEvents(room: Room) {
+    await this.subscriber.psubscribe(`__keyspace@0__:${room}:*`);
+    await this.subscriber.subscribe(`${room}`);
+  }
+
+  private makeSureRoomIsJoined(room: Room) {
+    if (!this.subscriptions.has(room)) {
+      throw new Error(`You are not joined in ${room}!`);
+    }
+  }
+
+  private async setupCallbacksIfNecessary() {
+    if (this.hasSetupCallbacks) {
+      return;
+    }
+
+    await this.subscriber.config('SET', 'notify-keyspace-events', 'Kx');
+    this.subscriber.on('pmessage', this.onExpiredKeyMessage);
+    this.subscriber.on('message', this.onPubSubMessage);
+
+    this.hasSetupCallbacks = true;
+  }
+
+  private onExpiredKeyMessage = (pattern: string, key: string) => {
+    const [, room, instance] = key.split(':');
+    const callback = this.subscriptions.get(room);
+    if (callback) {
+      callback({
+        type: MessageTypes.Leave,
+        sender: instance,
+      });
+    }
+  }
+
+  private onPubSubMessage = (channel: string, payload: string) => {
+    const callback = this.subscriptions.get(channel);
+    if (callback) {
+      const message = this.deserialize(payload);
+      if (message.sender === this.instance) {
+        return;
+      }
+
+      callback(message);
+    }
+  }
+
+  private serialize(message: Message) {
+    return JSON.stringify(message);
+  }
+
+  private deserialize(payload: string) {
+    return JSON.parse(payload) as Message;
```

Comment:
Oops, didn't see the cast here. It's unsafe, though - perhaps use a typeguard instead of a hard cast?

Reply:
I don't think we have to, it's the kinda the same discussion as to if we need to check if the `callback` is defined. If we trust the data from redis then we know for sure that eveyrhing that comes was already compile time validated before it was sent.
If not, then we have to revert the check if th `callback` exists, if the `payload` is a valid json and if the parsed message is actually a type safe Message.

Reply:
In the end, this is a hard cast, which can be avoided - in my view, any hard casting is dangerous. When the language provides a way to handle this gracefully, I don't see why not.

With regards to the callback, we're adhering to the interface of the `Map` class and its `get` method. Here, we're  not  doing anything to ensure we're actually returning a `Message`. This can easily lead to hidden problems at some point.

Reply:
The language encourages you to validate the input runtime and consider it type safe onwards, that's what I have done here keeping in mind the assumption that redis is not an external system and we can assume that anything that we receive from it is sent by us., hence the usage of hard casts.
The current approach can be considered more fragile, any developer that works on it will have to understand the implications of potentially publishing malformed data in first place. On the other hand, the other approach of runtime validating eveything that TS can't grasp simply to avoid `as` is overly defensive, it's no longer clear when something is actual runtime validation of the input and when it is done merely to satisy the compiler.

Reply:
Ofc, ^ is all based on my assumption that data received from redis is not considered potentially malformed.

Reply:
> Ofc, ^ is all based on my assumption that data received from redis is not considered potentially malformed.

Which is a real possibility.

Effectively, we have a system boundary here, as we're listening on stuff coming from an external process. At the moment, we assume we know what we'll be getting - we should validate it.

Reply:
I am okay with changing my original assumption and threating anything receive from redis as potentially malformed.

Reply:
Code is now assuming redis is external.